### ISSUE TITLE
Creating a connection

### ISSUE THREAD

**Issue by Anas-1998** [tkn-tub/serpentine-env#4]
Hello, Thanks for this great work.
It has been really difficult for me to build the connection for our project as the documentation is very simplified, and the example is a bit complicated and uses extra material.
I would really appreciate it if you just specify what files we simply need to include to our project to assign the action, reward and observation. and what files we need to build the connection through the proto.
Thanks again

**Comment by dbuse** [comment]
Hi @Anas-1998 
could you please take a look at the other Veins-Gym Environment we published: [DCC Env](https://github.com/tkn-tub/dcc-env). It is much simpler and does not require Veins-VLC or other extra material. So I hope that makes the process of creating your own environment easier.
Also, there is a writeup on this on the [docs of the Veins-Gym repository](https://github.com/tkn-tub/veins-gym/blob/master/doc/getting_started.md), make sure to check that out as well, if you have not done so, yet.

**Comment by lionyouko** [comment]
I was about to ask the same. 
I will check the listed material.

I have a very simple network using omnet++ only (but it will use veins further ahead). 

To make the POC to my teacher, I need to link from omnet++ up to openAI gym, but the initial documentation was like Anas said (complex example, yet simple explanation), which was confusing me. 

I know you already kind for producing this nice middleware, we thank you for that, but documentation is being hard to follow a bit (since yesterday I am trying to put it together without the extras, still no concrete result).

Thank you.

**Comment by Anas-1998** [comment]
Hi @lionyouko 
I hope you have done it already. If you did not, let me give you some hints since I have wasted so much time on this and it has just worked for me!.
First of all, run veins independently, so if you are getting any progress/errors, you can see that in the command window!
to do that, you can check out this link: 
https://github.com/tkn-tub/veins-gym
down you will see how to do that
Secondly, this is what basically you need:
The gymConnection.Ned, gymConnection.cc and gymConnection.h files from either environments, just copy paste them to your environment
Thirdly, go to wherever you want to send your observation space and reward, and receive the action from RL: in that file, you do this:
in the h file:
add: 
```
#include "packageName/GymConnection.h"
Public:
virtual ~ClassName();
Protected:
GymConnection *gymCon = nullptr;
Private:
veinsgym::proto::Request serializeObservation(const std::array<double, 4> &observation, double reward) const;
// 4 is just the observation size, for simplicity, try to send the same as the original file (4) values.
```
in the cc file:
```
//Include the headerfile for sure! and then write this before anything, this is for the shutdown I believe
ClassName::~ClassName(){
if (gymCon) {
        veinsgym::proto::Request request;
        request.set_id(1);
		*(request.mutable_shutdown()) = {};
		auto response = gymCon->communicate(request);
    }
}
// In the initialize function dont forget to add any constrains for when you want the connection
//after passing the constrains write this:
gymCon = veins::FindModule<GymConnection*>::findGlobalModule();
        ASSERT(gymCon);

and then whenever u need an action, you should send observation and reward like this:
gymCon = veins::FindModule<GymConnection*>::findGlobalModule();
        ASSERT(gymCon); // try to send just some test constant values
        const auto observation = test_observation;
        const auto reward = test_reward;
        const auto request = serializeObservation(observation, reward);
        auto response = gymCon->communicate(request);
        auto result = response.action().discrete().value();
// Dont forget to add this function in the cc file:
veinsgym::proto::Request SerpentineApp::serializeObservation(const std::array<double, 4> &observation, const double reward) const {
    veinsgym::proto::Request request;
    request.set_id(1);
    auto *values = request.mutable_step()->mutable_observation()->mutable_box()->mutable_values();
    *values = {observation.begin(), observation.end()};
    request.mutable_step()->mutable_reward()->mutable_box()->mutable_values()->Add();
    request.mutable_step()->mutable_reward()->mutable_box()->set_values(0, reward);
    return request;
}
```
now save this and remember the following notes, I dont know if they are all critical or not, maybe the authors can confirm, but for me I just did whatever they did in their environment and it finally worked for me
So remember this:
1- if you changed observation size, remember to change minimum and maximum values in the ini file
2- in the main ned file, they used managerForker, so I switched to that aswell instead of the normal Manager
3- add gymcon module to the main ned file as in their example!
I hope I did not forget anything and I hope it works for you as well. Good Luck!

**Comment by lionyouko** [comment]
Hi, @Anas-1998 , first of all, thank you. I will read these directions carefully and I come back to say if it worked for me. I hope It works, I am pretty much depending on this tool to be able to keep going with Omnet++. 

@dbuse I hope Mr Buse is doing well. I would like to humbly suggest Mr Buse could make a tutorial pretty much like this one from https://github.com/tkn-tub/ns3-gym/blob/app/README.md#Examples and https://github.com/hust-diangroup/ns3-ai. 

I think Anas may have helped me in huge way. I will try her directions by morning (in my timezone). 
Thank you.

**Comment by dbuse** [comment]
Thanks, @Anas-1998 , for your explanations. Regarding your notes / questions:

> 1- if you changed observation size, remember to change minimum and maximum values in the ini file

Yes, the observation configuration (in the `omnetpp.ini` file) has to be adapted for your shape and minimum/maximum values.

> 2- in the main ned file, they used managerForker, so I switched to that aswell instead of the normal Manager

Yes, we recommend using the `TraCIScenarioManagerForker`. In contrast to the `TraCIScenarioManagerLaunchd`, it does not require the launchd script to run in parallel, removing another moving part from the system.

> 3- add gymcon module to the main ned file as in their example!

Yes, this is essential. The `GymConnection` has to live somewhere in the simulation. With this approach, it is simply inserted into the Network as a top-level module. This also ensures that there is exactly one instance of it. Then it can be found and accessed from any place in the simulation via `veins::FindModule<GymConnection*>::findGlobalModule()`.

We'll take your notes and suggestions into consideration for future updates to the documentation. If you publish your own Gym environments, e.g., here on GitHub, please let us know so we can link them as further examples.

**Comment by lionyouko** [comment]
Hey, sorry to open this again. I would like to ask if none of you had problem with protobuf. 

When I am compiling the project in Pmnet++ IDE, it gives more than 270 errors of the type undefined reference regarding the classes and methods of protobuffer. I tried lots of stuff, but still have this undefined reference problems. 

I will let a link that talks about it, but I really don't know how to put it to work correctly. 

https://stackoverflow.com/questions/30124264/undefined-reference-to-googleprotobufinternalempty-string-abicxx11

it says something regarding ABI issues, and I never seem this before.

If I run serpentine-env and dcc-app, I can run them from snakemake and make, but for my project the compiled from veingym.proto is giving me such undefined reference for proto::google::<class> errors (and I found the classes already inside protobuf google dir)

SOPrry for bothering again

**Comment by lionyouko** [comment]
After 10h literally, I made the project run, however it is not connecting to the agent. It Seems it is because of them run file. @Anas-1998 , did u altered something in that file? Also: did u create src/experiment and such dirs to be able to run? All my files related to the omnet++ simulation are in root dir of the project, I don't know if this is getting in the way. 

Python agent tries to connect for 5 seconds and then crash because no response. 

I have gym con in .ned and I initialize it in my agent sender, the class that is supposed to communicate with python agent. 

Thanks again
[agentsender.txt](https://github.com/tkn-tub/serpentine-env/files/8374978/agentsender.txt)

**Comment by dbuse** [comment]
I'm not using the OMNeT++ IDE (the environment based on Eclipse), so I can't really help with build issues on that. What is important though, is that if you run build commands from both the commandline (snakemake or make) and the IDE, you have to make extra sure that the compiler arguments are the same. E.g., the Stackoverflow Link you mentioned talks about a change that results from using the C++11 standard (which Veins and Veinsgym require) and previous versions. As far as I can tell, this leads the to the ABI issues.

Regarding the other issue: try running the veins simulation separately and with a debugger. See [here](https://github.com/tkn-tub/veins-gym#starting-veins-simulations-separately), but use `./run -d -t gdb -u Cmdenv '--*.gym-connection.port=5555'` (the `-d` runs the debug version and `-t gdb` runs veins within a debugger). That can hopefully give you some further hints.
